File: src/main/java/xyz/mizarc/persistentitems/commands/PISubCommands/ShowCommand.java
package xyz.mizarc.persistentitems.commands.PISubCommands;

import org.bukkit.Bukkit;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import xyz.mizarc.persistentitems.DatabaseConnection;
import xyz.mizarc.persistentitems.Item;
import xyz.mizarc.persistentitems.ItemContainer;
import xyz.mizarc.persistentitems.PersistentItems;
import xyz.mizarc.persistentitems.commands.SubCommand;

import java.util.UUID;

public class ShowCommand implements SubCommand {
    PersistentItems plugin;

    public ShowCommand(PersistentItems plugin) {
        this.plugin = plugin;
    }

    @Override
    public boolean execute(CommandSender sender, String[] args) {
        if (args.length == 0) {
            sender.sendMessage("You must specify the name of an item");
            return false;

        } else if (args.length == 1) {
            if (!(sender instanceof Player)) {
                sender.sendMessage("You must be a player to run this command without the player argument");
                return false;
            }

            if (!isItemExists(args[0])) {
                sender.sendMessage("That item does not exist");
                return false;
            }

            Player player = (Player) sender;
            removeEntry(player, args[0]);
            addToInventory(player, args[0]);
            return true;
        }

        if (!isItemExists(args[0])) {
            sender.sendMessage("That item does not exist");
            return false;
        }


        if (Bukkit.getServer().getPlayer(args[1]) == null) {
            sender.sendMessage("That player is not online");
            return false;
        }

        Player player = (Player) sender;
        removeEntry(player, args[0]);
        addToInventory(player, args[0]);
        return true;
    }

    private boolean isItemExists(String string) {
        ItemContainer container = plugin.getItemContainer();
        return container.getItem(string) != null;
    }

    private void removeEntry(Player player, String itemName) {
        DatabaseConnection database = new DatabaseConnection(plugin);
        if (!database.isHidden(player.getUniqueId().toString(), itemName, "global")) {
            database.removeHidden(player.getUniqueId().toString(), itemName, "global");
        }
    }

    private void addToInventory(Player player, String itemName) {
        ItemContainer container = plugin.getItemContainer();
        Item item = container.getItem(itemName);
        ItemStack itemStack = item.getItemStack(plugin);

        if (!player.getInventory().contains(itemStack)) {
            Inventory inventory = player.getInventory();

            if (inventory.getItem(item.getSlot()) == null) {
                inventory.setItem(item.getSlot(), itemStack);
                return;
            }
            inventory.addItem(itemStack);
        }
    }
}

File: src/main/kotlin/dev/mizarc/itembinds/commands/itembinds/ShowCommand.kt
package dev.mizarc.itembinds.commands.itembinds;

import co.aikar.commands.ConditionFailedException;
import co.aikar.commands.MessageKeys;
import co.aikar.commands.annotation.*;
import co.aikar.commands.annotation.Optional
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.PlayerInventory;
import dev.mizarc.itembinds.Item;
import dev.mizarc.itembinds.commands.ItemBindsCommand;
import java.util.*

@CommandAlias("itembinds")
class ShowCommand: ItemBindsCommand() {

    @Subcommand("show")
    @CommandPermission("itembinds.command.show")
    @CommandCompletion("@pitems @players")
    @Syntax("<item> [player]")
    fun onShow(sender: CommandSender, name: String) {
        val item = itemRepo.getByName(name).firstOrNull()

        // Error if persistent item is not active
        if (item == null) {
            sender.sendMessage("Item $name does not exist");
            return;
        }

        // Error if console is trying to use this without specifying a player
        if (sender !is Player) {
            sender.sendMessage("You must specify the player argument as the console");
            return;
        }

        //Add item to own inventory unless you already have it
        if (!addToInventory(sender.inventory, item.id)) {
            sender.sendMessage("Item ${item.name} is already in your inventory");
            return;
        }

        playerItemsRepo.remove(sender, item)
        sender.sendMessage("Item ${item.name} has been added to your inventory");
    }

    fun onShowOthers(sender: CommandSender, item: Item, player: Player) {
        // Error if player doesn't have the 'others' permission
        if (!sender.hasPermission("persistentitems.command.show.others")) {
            throw ConditionFailedException(MessageKeys.PERMISSION_DENIED_PARAMETER);
        }

        // Add item to specified player's inventory unless player already has it
        if (!addToInventory(player.inventory, item.id)) {
            sender.sendMessage("Item ${item.name} is already in ${player.displayName()}'s inventory")
            return;
        }
        playerItemsRepo.remove(player, item)
        sender.sendMessage("Item ${item.name} has been added to ${player.displayName()}'s inventory")
    }

    private fun addToInventory(inventory: PlayerInventory, itemId: UUID): Boolean {
        val presentItems: List<ItemStack> = persistencyService.getItemsInInventory(inventory, itemId);
        if (presentItems.isNotEmpty()) {
            return false;
        }

        // True if the inventory doesn't have the item and attempt to add the item to the specified slot.
        // Add to any slot if that slot is occupied
        val item = itemRepo.getById(itemId) ?: return false
        if (inventory.getItem(item.slot) == null) {
            inventory.setItem(item.slot, item.itemStack);
            return true
        }
        inventory.addItem(item.itemStack);
        return true
    }
}
